File: shell/gnome/src/prefs/accountRow.js
import Adw from 'gi://Adw';
import Gtk from 'gi://Gtk';
import { percentLeft, windowLabel } from '../format.js';
import { _ } from '../i18n.js';
import { accountName } from '../providers.js';
import { hiddenWindowsAfter, hiddenWindowsPatch, isWindowHidden } from '../settings.js';
import { signInTarget } from './registry.js';
import { providerImage } from './widgets.js';

function subtitleOf(account) {
    const parts = [account.providerName, account.plan, account.label ? account.email : null];
    if (account.owner === 'headroom') parts.push(_('added in Headroom'));
    return parts.filter(Boolean).join(' · ');
}

function removeSubtitle(account) {
    if (account.owner === 'headroom') return _('Deletes the sign-in Headroom created for this account');
    return _('Stops showing this account. Its CLI stays signed in.');
}

function shapeOf(account) {
    return JSON.stringify([account.owner, account.windows.map(window => [window.id, window.label])]);
}

function iconButton(iconName, tooltip, onClick) {
    const button = new Gtk.Button({ icon_name: iconName, tooltip_text: tooltip, valign: Gtk.Align.CENTER });
    button.connect('clicked', () => onClick());
    return button;
}

export class AccountRow {
    constructor({ account, dir, client, onMove, onRemove, onSignIn }) {
        this._client = client;
        this._actions = { onMove, onRemove, onSignIn };
        this._syncing = false;
        this._children = [];
        this.id = account.id;
        this.widget = new Adw.ExpanderRow({ use_markup: false });
        const handle = new Gtk.Image({ icon_name: 'list-drag-handle-symbolic', css_classes: ['dim-label'] });
        this.widget.add_prefix(handle);
        this.widget.add_prefix(providerImage(dir, account.provider, 24));
        this._signIn = new Gtk.Button({ label: _('Sign in again…'), valign: Gtk.Align.CENTER, visible: false });
        this._signIn.connect('clicked', () => this._actions.onSignIn(this._account));
        this.widget.add_suffix(this._signIn);
        this._visible = new Gtk.Switch({ valign: Gtk.Align.CENTER, tooltip_text: _('Show in the panel and popup') });
        this._visible.connect('notify::active', () => this._onVisibleToggled());
        this.widget.add_suffix(this._visible);
    }

    update(account, settings) {
        this._account = account;
        this._syncing = true;
        this.widget.title = accountName(account);
        this.widget.subtitle = subtitleOf(account);
        this._visible.active = !account.hidden;
        this._signIn.visible = signInTarget(account, this._client.providers) !== null;
        if (this._shape !== shapeOf(account)) this._rebuild(account);
        this._syncChildren(account, settings);
        this._syncing = false;
    }

    _onVisibleToggled() {
        if (!this._syncing) this._client.setAccountHidden(this.id, !this._visible.active);
    }

    _rebuild(account) {
        for (const child of this._children) this.widget.remove(child);
        this._labelRow = this._labelEntry();
        this._windowRows = account.windows.map(window => this._windowRow(window));
        this._children = [this._labelRow, ...this._windowRows, this._positionRow(), this._removeRow(account)];
        for (const child of this._children) this.widget.add_row(child);
        this._shape = shapeOf(account);
    }

    _syncChildren(account, settings) {
        const editing = (this._labelRow.get_state_flags() & Gtk.StateFlags.FOCUS_WITHIN) !== 0;
        if (!editing && this._labelRow.text !== (account.label ?? '')) this._labelRow.text = account.label ?? '';
        account.windows.forEach((window, index) => {
            const row = this._windowRows[index];
            row.active = !isWindowHidden(settings.display, account.id, window.id);
            row.subtitle = window.remainingPercent === null ? _('No data yet') : percentLeft(window.remainingPercent);
        });
    }

    _labelEntry() {
        const row = new Adw.EntryRow({ title: _('Label'), show_apply_button: true });
        row.connect('apply', () => this._client.setAccountLabel(this.id, row.text.trim()));
        return row;
    }

    _windowRow(window) {
        const row = new Adw.SwitchRow({ title: windowLabel(window.id, window.label), use_markup: false });
        row.connect('notify::active', () => {
            const display = this._client.settings?.display;
            if (this._syncing || !display) return;
            const windows = hiddenWindowsAfter(display, this.id, window.id, !row.active);
            this._client.updateSettings(hiddenWindowsPatch(this.id, windows));
        });
        return row;
    }

    _positionRow() {
        const row = new Adw.ActionRow({ title: _('Position') });
        const buttons = new Gtk.Box({ css_classes: ['linked'], valign: Gtk.Align.CENTER });
        buttons.append(iconButton('go-up-symbolic', _('Move up'), () => this._actions.onMove(this.id, -1)));
        buttons.append(iconButton('go-down-symbolic', _('Move down'), () => this._actions.onMove(this.id, 1)));
        row.add_suffix(buttons);
        return row;
    }

    _removeRow(account) {
        const row = new Adw.ActionRow({ title: _('Remove from Headroom'), subtitle: removeSubtitle(account) });
        const button = new Gtk.Button({ label: _('Remove'), valign: Gtk.Align.CENTER });
        button.add_css_class('destructive-action');
        button.connect('clicked', () => this._actions.onRemove(this._account));
        row.add_suffix(button);
        return row;
    }
}

File: crates/headroom-tray/src/ui/prefs/account_row.rs
use std::cell::{Cell, RefCell};
use std::rc::Rc;

use adw::prelude::*;

use super::{Act, PrefsAction};
use crate::assets::{CLAUDE_COLOR, Tint, provider_logo};
use crate::i18n::Lang;
use crate::payload::Account;
use crate::preferences::change::Change;
use crate::preferences::choices::{account_name, account_subtitle, removal_subtitle};
use crate::preferences::model::DisplaySettings;
use crate::ui::widgets::svg_image;

const LOGO_SIZE: i32 = 24;

pub type Mover = Rc<dyn Fn(&str, isize)>;
pub type Remover = Rc<dyn Fn(&Account)>;

pub struct AccountRow {
    pub widget: adw::ExpanderRow,
    pub id: String,
    pub shape: Vec<(String, String)>,
    visible: gtk::Switch,
    label: adw::EntryRow,
    windows: Vec<(String, adw::SwitchRow)>,
    syncing: Rc<Cell<bool>>,
    display: Rc<RefCell<DisplaySettings>>,
    account: Rc<RefCell<Account>>,
}

pub fn provider_image(provider: &str, text_color: &str, size: i32) -> gtk::Image {
    let (svg, tint) = provider_logo(provider);
    let color = match tint {
        Tint::Brand => CLAUDE_COLOR,
        Tint::Text => text_color,
    };
    svg_image(svg, color, size, &[])
}

fn icon_button(icon: &str, tooltip: &str, on_click: impl Fn() + 'static) -> gtk::Button {
    let button = gtk::Button::from_icon_name(icon);
    button.set_tooltip_text(Some(tooltip));
    button.set_valign(gtk::Align::Center);
    button.connect_clicked(move |_| on_click());
    button
}

#[must_use]
pub fn shape(account: &Account) -> Vec<(String, String)> {
    let head = (format!("{:?}", account.owner), account.provider.clone());
    std::iter::once(head)
        .chain(
            account
                .windows
                .iter()
                .map(|window| (window.id.clone(), window.label.clone())),
        )
        .collect()
}

fn window_rows(account: &Account) -> Vec<(String, adw::SwitchRow)> {
    account
        .windows
        .iter()
        .map(|window| {
            let row = adw::SwitchRow::builder()
                .title(&window.label)
                .use_markup(false)
                .build();
            (window.id.clone(), row)
        })
        .collect()
}

impl AccountRow {
    pub fn new(lang: Lang, account: &Account, logo_color: &str) -> Self {
        let widget = adw::ExpanderRow::builder().use_markup(false).build();
        widget.add_prefix(&provider_image(&account.provider, logo_color, LOGO_SIZE));
        let visible = gtk::Switch::new();
        visible.set_valign(gtk::Align::Center);
        visible.set_tooltip_text(Some(lang.tr("Show in the tray and popup")));
        widget.add_suffix(&visible);
        let label = adw::EntryRow::builder()
            .title(lang.tr("Label"))
            .show_apply_button(true)
            .build();
        Self {
            widget,
            id: account.id.clone(),
            shape: shape(account),
            visible,
            label,
            windows: window_rows(account),
            syncing: Rc::default(),
            display: Rc::default(),
            account: Rc::new(RefCell::new(account.clone())),
        }
    }

    pub fn assemble(&self, lang: Lang, act: &Act, mover: &Mover, remover: &Remover) {
        self.connect_visibility(act);
        self.connect_label(act);
        self.widget.add_row(&self.label);
        for (window_id, switch) in &self.windows {
            self.connect_window(act, window_id, switch);
            self.widget.add_row(switch);
        }
        self.widget.add_row(&self.position_row(lang, mover));
        self.widget.add_row(&self.remove_row(lang, remover));
    }

    fn connect_visibility(&self, act: &Act) {
        let (act, id, syncing) = (Rc::clone(act), self.id.clone(), Rc::clone(&self.syncing));
        self.visible.connect_active_notify(move |switch| {
            if !syncing.get() {
                act(PrefsAction::SetHidden {
                    account_id: id.clone(),
                    hidden: !switch.is_active(),
                });
            }
        });
    }

    fn connect_label(&self, act: &Act) {
        let (act, id) = (Rc::clone(act), self.id.clone());
        self.label.connect_apply(move |entry| {
            act(PrefsAction::SetLabel {
                account_id: id.clone(),
                label: entry.text().trim().to_owned(),
            });
        });
    }

    fn connect_window(&self, act: &Act, window_id: &str, switch: &adw::SwitchRow) {
        let act = Rc::clone(act);
        let (id, window) = (self.id.clone(), window_id.to_owned());
        let (syncing, display) = (Rc::clone(&self.syncing), Rc::clone(&self.display));
        switch.connect_active_notify(move |switch| {
            if syncing.get() {
                return;
            }
            let windows = display
                .borrow()
                .hidden_windows_after(&id, &window, !switch.is_active());
            act(PrefsAction::Change(Change::HiddenWindows {
                account_id: id.clone(),
                windows,
            }));
        });
    }

    fn position_row(&self, lang: Lang, mover: &Mover) -> adw::ActionRow {
        let row = adw::ActionRow::builder().title(lang.tr("Position")).build();
        let buttons = gtk::Box::new(gtk::Orientation::Horizontal, 0);
        buttons.add_css_class("linked");
        buttons.set_valign(gtk::Align::Center);
        for (icon, tooltip, delta) in [
            ("go-up-symbolic", lang.tr("Move up"), -1),
            ("go-down-symbolic", lang.tr("Move down"), 1),
        ] {
            let (mover, id) = (Rc::clone(mover), self.id.clone());
            buttons.append(&icon_button(icon, tooltip, move || mover(&id, delta)));
        }
        row.add_suffix(&buttons);
        row
    }

    fn remove_row(&self, lang: Lang, remover: &Remover) -> adw::ActionRow {
        let owner = self.account.borrow().owner;
        let row = adw::ActionRow::builder()
            .title(lang.tr("Remove from Headroom"))
            .subtitle(removal_subtitle(lang, owner))
            .build();
        let button = gtk::Button::with_label(lang.tr("Remove"));
        button.add_css_class("destructive-action");
        button.set_valign(gtk::Align::Center);
        let (remover, account) = (Rc::clone(remover), Rc::clone(&self.account));
        button.connect_clicked(move |_| remover(&account.borrow()));
        row.add_suffix(&button);
        row
    }

    pub fn update(&self, lang: Lang, account: &Account, display: &DisplaySettings) {
        self.syncing.set(true);
        self.widget.set_title(&account_name(account));
        self.widget.set_subtitle(&account_subtitle(lang, account));
        self.visible.set_active(!account.hidden);
        let editing = self
            .label
            .state_flags()
            .contains(gtk::StateFlags::FOCUS_WITHIN);
        let label = account.label.as_deref().unwrap_or_default();
        if !editing && self.label.text() != label {
            self.label.set_text(label);
        }
        for (window_id, switch) in &self.windows {
            switch.set_active(!display.is_window_hidden(&account.id, window_id));
        }
        *self.display.borrow_mut() = display.clone();
        *self.account.borrow_mut() = account.clone();
        self.syncing.set(false);
    }
}
